firmware/shared/src/hw/hw_usb.c

```diff
@@ -33,133 +37,143 @@ static const osMessageQueueAttr_t rx_queue_attr = { .name      = "USB RX Queue",
+    if (status != osOK)
     {
-        // Dump the byte.
-        osStatus_t status = osMessageQueueGet(rx_queue_id, &dest[i], NULL, osWaitForever);
-
-        // Check success.
-        if (status != osOK)
+        if (status == osErrorTimeout)
         {
-            LOG_WARN("USB: Queue pop returned non-ok status %d", status);
-            return false;
+            LOG_WARN("USB: Timeout occurred while waiting for RX queue.");
+            return EXIT_CODE_TIMEOUT;
         }
+        LOG_WARN("USB: Queue pop returned non-ok status %d", status);
+        return EXIT_CODE_ERROR;
     }
-
-    return true;
+    return EXIT_CODE_OK;
 }
 
-bool hw_usb_pushRxMsgToQueue(uint8_t *msg, uint32_t len)
+bool hw_usb_pushRxMsgToQueue(const uint8_t *msg, const uint32_t len)
 {
     if (rx_queue_id == NULL)
     {
         LOG_ERROR("USB: Peripheral not initialized before attempting push to RX queue.");
         return false;
     }
 
-    uint32_t space = osMessageQueueGetSpace(rx_queue_id);
+    const uint32_t space = osMessageQueueGetSpace(rx_queue_id);
     if (len > space)
     {
         LOG_ERROR("USB: Receiving message that will overflow RX queue.");
         return false;
     }
-
     for (uint32_t i = 0; i < len; i += 1)
     {
         // Note: timeout cannot be non-zero in an IRQ.
-        osStatus_t status = osMessageQueuePut(rx_queue_id, &msg[i], 0, 0);
+        const osStatus_t status = osMessageQueuePut(rx_queue_id, &msg[i], 0, 0);
         if (status != osOK)
         {
             LOG_ERROR(
                 "USB: Error status encountered when pushing to RX queue, "
-                "osMessageQueuePut returned %d insteado of osOK.",
+                "osMessageQueuePut returned %d instead of osOK.",
                 status);
             return false;
         }
     }
-
     return true;
 }
 
+// Connection handling
+
+static bool         usb_connected = false;
+static TaskHandle_t xTaskToNotify = NULL;
+
+void hw_usb_connect_callback()
+{
+    usb_connected = true;
+    if (xTaskToNotify == NULL)
+    {
+        LOG_WARN("USB: No task to notify.");
+        return;
+    }
+    BaseType_t higherPriorityTaskWoken = pdFALSE;
+    // notify the task which is waiting for the USB connection
+    const BaseType_t status = xTaskNotifyFromISR(xTaskToNotify, 0, 0, &higherPriorityTaskWoken);
+    assert(status == pdPASS);
+    portYIELD_FROM_ISR(higherPriorityTaskWoken);
+}
+
+void hw_usb_disconnect_callback()
+{
+    usb_connected = false;
+}
+
+bool hw_usb_connected()
+{
+    return usb_connected;
+}
+
+void hw_usb_waitForConnected()
+{
+    if (usb_connected)
+        return;
+
+    assert(xTaskToNotify == NULL);
+    // save which task we need to notify
+    xTaskToNotify = xTaskGetCurrentTaskHandle();
+    // block this task until notification comes
+    const BaseType_t status = xTaskNotifyWait(0, 0, NULL, portMAX_DELAY);
+    assert(status == pdPASS);
+    xTaskToNotify = NULL;
+}
```

Comment:
maybe stick with old api? is this one better?

Reply:
ok migrated